CIP-????/README.md

```diff
@@ -0,0 +1,320 @@
+    const rightChildIndex = 2 * index + 2
+
+    const valHash = hashVal(val)
+    const leftHash = this.hashNode(leftChildIndex)
+    const rightHash = this.hashNode(rightChildIndex)
+
+    return combineThreeHashes(valHash, leftHash, rightHash)
+  }
+```
+
+> Note: A compact array based structure is used for off-chain representation of the tree.
+
+## Minimal Subtree (Proof)
+
+The proof provided by the solution is a minimal subtree together with the required values (depends on the proofs). In other words, the minimal subtre is a path from the root node to the provable node (membership or non-membership) contains hash of the nodes' value the hash of the child not in the path down to the node and the child contains the node as path down to the provabel node.
+
+> Note: The minimal subtree is used for proofs and contains only hash values of Vals, nodes and branches along the path of specific nodes (e.g., membership, non-membership, update, delete).
+
+Typescript definition example of the minimal subtree (TreeProof)
+
+``` typescript
+export type TreeProof =
+  | {
+      // `hash` is the valHash of the Val of the node
+      HashNode: { hash: string; left: TreeProof; right: TreeProof }
+    }
+  | {
+      // `hash` is the rootHash of a node, a branch etc.
+      NodeHash: { hash: string }
+    }
+```
+
+
+This structure regenerates the exact hash of the tree as the whole IntegriTree would be hashed.
+
+In merkle tree the leaf pairs (that contain data) are hashed up to the tree while in IntegriTree the nodes' values (`Val`) are part of the proof.
+For example, for the following Proof
+
+``` typescript
+const proof = HashNode {
+  hash: valHash(Val{ xi: 1, xa: "a", xb: "adam"}),
+  left: HashNode {
+            hash: valHash(Val{ xi: 2, xa: "adam", xb: "b"}),
+            left: emptyHash,
+            right: emptyHash
+        },
+  right: emptyHash
+}
+```
+
+It's easily can be proven that "adam" is in the tree, but "aby" is not, having a `Val { xi: "1", xa: "a", xb: "adam"}` and the Proof
+as the proof's `rootHash(proof)` exactly the same with the rootHash of the whole integritree, and the hash of proof contains the hash of the required proovable `Val`
+
+### Root Hash Calculation of the minimal subtree
+
+The root hash of the minimal subtree use similar function as in the IntegriTree, see the `aiken`'s implementation below:
+
+``` gleam
+pub fn root_hash(root: Proof) -> Hash {
+  when root is {
+    NodeHash { hash } -> hash
+    HashNode { hash, left, right } ->
+      combine_three_hashes(hash, root_hash(left), root_hash(right))
+  }
+}
+```
+
+## Membership and Non-Membership Proofs
+
+IntegriTree provides proofs for both membership and non-membership of elements in the tree:
+
+- **Membership**: If an `x` element is in the tree, there exists at least one node where either `xa = x` or `xb = x`.
+- **Non-Membership**: If an element is not in the tree, there exists a node where `xa < x < xb`.
+
+## Append and Delete Proofs
```

Comment:
```suggestion
### Append and Delete Proofs
```

Reply:
@ilap very important to maintain standard document sections.